From 2d965000a24e450cb5e55464bdb45f66e2bb9b86 Mon Sep 17 00:00:00 2001
From: Reuben Kuhnert <reuben.kuhnert@cloudera.com>
Date: Tue, 24 May 2016 06:44:48 -0500
Subject: [PATCH 0606/1164] CLOUDERA-BUILD: CDH-40690: [test regression]
 HIVE-13696

FairSchedulerShim update does not check if the username is 'null'. Adds guard to ensure that NPEs don't
occur when this is the case. The upstream patch for HIVE-13696 will also be updated to contain the fix for this issue.

Change-Id: I32af7bd2e3734273b7d1b1546b8b19ba87488816
---
 ql/src/java/org/apache/hadoop/hive/ql/Driver.java  |    1 +
 .../hadoop/hive/schshim/FairSchedulerShim.java     |    8 ++++++--
 2 files changed, 7 insertions(+), 2 deletions(-)

diff --git a/ql/src/java/org/apache/hadoop/hive/ql/Driver.java b/ql/src/java/org/apache/hadoop/hive/ql/Driver.java
index 73e4d40..fcbba5e 100644
--- a/ql/src/java/org/apache/hadoop/hive/ql/Driver.java
+++ b/ql/src/java/org/apache/hadoop/hive/ql/Driver.java
@@ -541,6 +541,7 @@ public int compile(String command, boolean resetTaskIds) {
   }
 
   private HiveConf configureScheduling(HiveConf configuration, String forUser) throws IOException, HiveException {
+    if (configuration == null || StringUtils.isEmpty(forUser)) return configuration;
     if (YarnFairScheduling.usingNonImpersonationModeWithFairScheduling(configuration)) {
         YarnFairScheduling.validateYarnQueue(configuration, forUser);
     }
diff --git a/shims/scheduler/src/main/java/org/apache/hadoop/hive/schshim/FairSchedulerShim.java b/shims/scheduler/src/main/java/org/apache/hadoop/hive/schshim/FairSchedulerShim.java
index 741d992..285b73b 100644
--- a/shims/scheduler/src/main/java/org/apache/hadoop/hive/schshim/FairSchedulerShim.java
+++ b/shims/scheduler/src/main/java/org/apache/hadoop/hive/schshim/FairSchedulerShim.java
@@ -98,8 +98,12 @@ private void setJobQueueForUserInternal(Configuration conf, String queueName, St
         LOG.info("Setting queue name to: '{}' for user '{}'", requestedQueue, forUser);
         conf.set(MR2_JOB_QUEUE_PROPERTY, requestedQueue);
       } else {
-        LOG.warn("Unable to set queue: {} for user: {}, resetting to user's default queue.", requestedQueue, forUser);
-        conf.set(MR2_JOB_QUEUE_PROPERTY, queuePolicy.assignAppToQueue(YarnConfiguration.DEFAULT_QUEUE_NAME, forUser));
+        requestedQueue = queuePolicy.assignAppToQueue(YarnConfiguration.DEFAULT_QUEUE_NAME, forUser);
+        if (StringUtils.isNotBlank(requestedQueue)) {
+          LOG.warn("Unable to set queue: {} for user: {}, resetting to user's default queue.", requestedQueue, forUser);
+        } else {
+          LOG.warn("Unable to set queue for: {}", forUser);
+        }
       }
     }
   }
-- 
1.7.9.5

